packages/iov-stream/src/reducer.ts

```diff
@@ -34,12 +34,20 @@ export class Reducer<T, U> {
     this.reducer = reducer;
     this.state = initState;
     this.completed = new Promise<void>((resolve, reject) => {
-      this.stream.addListener({
-        complete: () => resolve(),
-        error: (err: any) => reject(err),
+      const subscription = this.stream.subscribe({
         next: (evt: T) => {
           this.state = this.reducer(this.state, evt);
         },
+        complete: () => {
+          resolve();
+          // this must happen after resolve, to ensure stream.subscribe() has finished
+          subscription.unsubscribe();
```

Comment:
I don't think the order is correct here. Caller code will continue to run after `resolve();` and it is unclear if and when `subscription.unsubscribe();` is executed.

You can safely unsubscribe first, since `stream.subscribe()` is done at the point where one of the `next`, `complete`, `error` callbacks are running.

Reply:
If you switch the error a number of tests fail inside of iov-stream.
I had it the other way at first